ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -748,6 +736,7 @@ void wasSuccessful() {
      * @throws Exception
      */
     @Override public void close() throws Exception {
+      shutdownHeartbeater();
```

Comment:
Theoretically this have the same issue as before the patch, just the other way around. We stop the heartbeat, the transaction times out, and we try to commit / abort.
How complicated would it be to turn off exception handling in the heart beater instead first, and stop it after abort / commit?

Would it worth the extra complexity?

Reply:
"Theoretically this have the same issue as before the patch, just the other way around. We stop the heartbeat, the transaction times out, and we try to commit / abort."
-> That's true, this was my first thought as well. However, I think shutting down the heartbeater should be really fast and not cause problems in healthy systems. If it's waiting to be scheduled by the executor, it will be shut down immediately. Otherwise it'll do one more heartbeating, but that heartbeating would need to take minutes (in line with the value of `hive.txn.timeout`) to cause any problems. If the heartbeating takes that long then we have other issues in the system anyway.

"How complicated would it be to turn off exception handling in the heart beater instead first, and stop it after abort / commit?"
-> Can you elaborate on what you mean by turning off exception handling? I think the general problem would remain: we commit/abort the txn and then send a signal to the heartbeater thread to stop doing whatever it's currently doing, but if it has already called the `msc.heartbeat()` method by that point (or it's just about to call it), there's nothing much we can do and it will lead to failure.

Reply:
We might just say to the heartbeater to ignore Exceptions at the shutdown phase

Reply:
As discussed, I've introduced a flag in the heartbeater thread to turn on/off error logging. The new order is now:
- turn off heartbeater error logging
- commit/abort txn
- shut down heartbeater

Reply:
Could I get an explanation too?:) What happens if the heartbeater is running, say it's on line 260, and the txn is committed – aren't we in our original boat then?

Reply:
Prior to calling commit(), we always set shouldLog=false. If the heartbeater is on 260 after commit has happened, the hms call will still fail but the error won't be logged on the HS2 side. However, the error would still be logged on the HMS side, which is not ideal. Flipping the order to first shut down the heartbeater and then commit/abort would be the only sure solution I think to avoid the HMS-side error log too.

Reply:
It's worse if the txn gets timed out (because then it's aborted). But I see your point that the race condition issue might still not be fixed with this change

Reply:
Yep, exactly. That's why we eventually chickened out out reordering the commit and the shutdown :D